include/javabind/codegen.hpp

```diff
@@ -0,0 +1,168 @@
+    {
+        os << "public enum " << class_name << " {\n";
+
+        for (std::size_t i = 0; i < binding.names().size(); ++i)
+        {
+            os << indent << binding.names().at(i);
+            if (i != binding.names().size() - 1) os << ",";
+            os << "\n";
+        }
+        os << "}\n";
+    }
+
+    static void write_record_class(std::ostream& os, std::string_view class_name, const std::vector<javabind::FieldBinding>& bindings)
+    {
+        os << "public record " << class_name << "(\n";
+
+        for (std::size_t i = 0; i < bindings.size(); ++i)
+        {
+            os << indent << bindings.at(i).type << " " << bindings.at(i).name;
+            if (i != bindings.size() - 1) os << ",\n";
+        }
+        os << ") {\n";
+        os << "}\n";
+    }
+
+    static void write_native_class(std::ostream& os, std::string_view class_name, const std::vector<javabind::FunctionBinding>& bindings)
+    {
+        auto is_close_method = [](const auto& binding) { return binding.name == "close" && binding.signature == FunctionTraits<void()>::sig; };
+        auto extends_native_object = std::any_of(bindings.begin(), bindings.end(), is_close_method);
+
+        if (extends_native_object)
+        {
+            os << "import hu.info.hunyadi.javabind.NativeObject;\n";
+            os << "\n";
+            os << "public class " << class_name << " extends NativeObject {\n";
+        }
+        else
+        {
+            os << "public class " << class_name << " {\n";
+        }
+
+        for (auto&& binding : bindings)
+        {
+            if (!binding.is_member)
+            {
+                os << indent << "public static native " << binding.return_display << " " << binding.name << "(" << binding.param_display << ");\n";
+            }
+        }
+        for (auto&& binding : bindings)
+        {
+            if (binding.is_member)
+            {
+                os << indent << "public native " << binding.return_display << " " << binding.name << "(" << binding.param_display << ");\n";
+            }
+        }
+        os << "}\n";
+    }
+
+    template<typename ContentWriter>
+    static void write_class(const std::filesystem::path& output_dir, const ClassDescription& class_desc, ContentWriter writer)
+    {
+        std::filesystem::path output_path = output_dir / class_desc.package_path;
+        std::filesystem::path output_filename = output_path / (class_desc.name + ".java");
+        std::filesystem::create_directories(output_path);
+        std::ofstream os { output_filename };
+
+        if (!os)
+        {
+            std::cerr << "Failed to open file: " << output_filename << std::endl;
+            return;
+        }
+        os << "package " << class_desc.package_name << ";\n";
+        os << "\n";
+        writer(os, class_desc.name);
+        os.close();
+    }
+
+    void codegen(const std::filesystem::path& output_path)
```

Comment:
This looks awesome. I am wondering if we could completely remove `print_registered_bindings`, and/or structure the function in a way that the two implementations share a common code base? Superficially, the difference seems that `print_registered_bindings` writes to the Java console, while `codegen` writes to a file stream. (Obviously, `codegen` looks more comprehensive.)

Reply:
I updated print_registered_bindings() to reuse code from the codegen header